integrations/helm/releasesync/releasesync.go

```diff
@@ -0,0 +1,349 @@
+/*
+When a Chart release is manually deleted/upgraded/created, the cluster gets out of sync
+with the prescribed state defined by FluxHelmRelease custom resources. The releasesync package
+attemps to bring the cluster back to the prescribed state.
+*/
+package releasesync
+
+import (
+	"context"
+	"fmt"
+	"strings"
+	"sync"
+	"time"
+
+	protobuf "github.com/golang/protobuf/ptypes/timestamp"
+	"github.com/weaveworks/flux/integrations/helm/chartsync"
+	"github.com/weaveworks/flux/integrations/helm/release"
+	"k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/client-go/kubernetes"
+	hapi_release "k8s.io/helm/pkg/proto/hapi/release"
+
+	"github.com/go-kit/kit/log"
+
+	ifv1 "github.com/weaveworks/flux/apis/helm.integrations.flux.weave.works/v1alpha"
+	ifclientset "github.com/weaveworks/flux/integrations/client/clientset/versioned" // kubernetes 1.9
+	helmgit "github.com/weaveworks/flux/integrations/helm/git"
+	chartrelease "github.com/weaveworks/flux/integrations/helm/release"
+)
+
+type action string
+
+const (
+	CustomResourceKind        = "FluxHelmRelease"
+	releaseLagTime            = 30
+	deleteAction       action = "DELETE"
+	installAction      action = "CREATE"
+	upgradeAction      action = "UPDATE"
+)
+
+type ReleaseFhr struct {
+	RelName string
+	FhrName string
+	Fhr     ifv1.FluxHelmRelease
+}
+
+type ReleaseChangeSync struct {
+	logger log.Logger
+	chartsync.Polling
+	kubeClient kubernetes.Clientset
+	ifClient   ifclientset.Clientset
+	release    *chartrelease.Release
+}
+
+func New(
+	logger log.Logger, syncInterval time.Duration, syncTimeout time.Duration,
+	kubeClient kubernetes.Clientset, ifClient ifclientset.Clientset,
+	release *chartrelease.Release) *ReleaseChangeSync {
+
+	return &ReleaseChangeSync{
+		logger:     logger,
+		Polling:    chartsync.Polling{Interval: syncInterval, Timeout: syncTimeout},
+		kubeClient: kubeClient,
+		ifClient:   ifClient,
+		release:    release,
+	}
+}
+
+type customResourceInfo struct {
+	name, releaseName string
+	resource          ifv1.FluxHelmRelease
+	lastUpdated       protobuf.Timestamp
+}
+
+type chartRelease struct {
+	releaseName  string
+	action       action
+	desiredState ifv1.FluxHelmRelease
+}
+
+// Run ... creates a syncing loop monitoring repo chart changes
+func (rs *ReleaseChangeSync) Run(stopCh <-chan struct{}, errc chan error, wg *sync.WaitGroup) {
+	rs.logger.Log("info", "Starting repo charts sync loop")
+
+	wg.Add(1)
+	go func() {
+		defer runtime.HandleCrash()
+		defer wg.Done()
+		defer rs.release.Repo.ReleasesSync.Cleanup()
```

Comment:
The Cleanup() works on a dir created when help-operator starts and clones a repo. It uses os.RemoveAll that returns nil when the path does not exist, so I thought that it is fine to use in case of a non existent dir. To avoid a situation where more processes are using the dir, I tried to use locking.
It did cross my mind before that this step is not really necessary -> Removed Cleanup()